core/system/src/ipc.c

```diff
@@ -288,6 +297,16 @@ void ipc_drain_queue(void)
             continue; /* Try the next send IO. */
         }
 
+        if (!uvisor_pool_queue_is_sane(recv_queue,
+                                       recv_queue->pool,
+                                       recv_array,
+                                       sizeof(*recv_array),
+                                       UVISOR_IPC_RECV_SLOTS)) {
+            /* This shouldn't happen in a non-malicious box. */
```

Comment:
Consider locking around the outside of the check, instead of inside. When locking inside, this _can_ happen in a non-malicious box.